wrapper/src/main/java/software/amazon/jdbc/hostlistprovider/monitoring/ClusterTopologyMonitorImpl.java

```diff
@@ -0,0 +1,851 @@
+        .forEach(
+            p -> {
+              this.monitoringProperties.put(
+                  p.substring(MONITORING_PROPERTY_PREFIX.length()),
+                  this.properties.getProperty(p));
+              this.monitoringProperties.remove(p);
+            });
+
+    // Set default values if they are not provided.
+    if (PropertyDefinition.SOCKET_TIMEOUT.getString(this.monitoringProperties) == null) {
+      PropertyDefinition.SOCKET_TIMEOUT.set(
+          this.monitoringProperties, String.valueOf(defaultSocketTimeoutMs));
+    }
+    if (PropertyDefinition.CONNECT_TIMEOUT.getString(this.monitoringProperties) == null) {
+      PropertyDefinition.CONNECT_TIMEOUT.set(
+          this.monitoringProperties, String.valueOf(defaultConnectionTimeoutMs));
+    }
+
+    this.monitorExecutor.submit(this);
+    this.monitorExecutor.shutdown(); // No more tasks are accepted by the pool.
+  }
+
+  @Override
+  public boolean canDispose() {
+    return true;
+  }
+
+  @Override
+  public void setClusterId(String clusterId) {
+    this.clusterId = clusterId;
+  }
+
+  @Override
+  public List<HostSpec> forceRefresh(final boolean writerImportant, final long timeoutMs)
+      throws SQLException, TimeoutException {
+
+    if (this.ignoreNewFailoverRequestsEndTimeNano.get() > 0
+        && System.nanoTime() < this.ignoreNewFailoverRequestsEndTimeNano.get()) {
+
+      // Previous failover has just completed. We can use results of it without triggering a new failover.
+      List<HostSpec> currentHosts = this.topologyMap.get(this.clusterId);
+      if (currentHosts != null) {
+        return currentHosts;
+      }
+    }
+
+    if (writerImportant) {
+      final Connection monitoringConnection = this.monitoringConnection.get();
+      this.monitoringConnection.set(null);
+      this.isVerifiedWriterConnection = false;
+      this.closeConnection(monitoringConnection, true);
+    }
+
+    return this.waitTillTopologyGetUpdated(timeoutMs);
+  }
+
+  @Override
+  public List<HostSpec> forceRefresh(@Nullable Connection connection, final long timeoutMs)
+      throws SQLException, TimeoutException {
+
+    if (this.isVerifiedWriterConnection) {
+      //push monitoring thread to refresh topology with a verified connection
+      return this.waitTillTopologyGetUpdated(timeoutMs);
+    }
+
+    // otherwise use provided unverified connection to update topology
+    return this.fetchTopologyAndUpdateCache(connection);
+  }
+
+  protected List<HostSpec> waitTillTopologyGetUpdated(final long timeoutMs) throws TimeoutException {
+
+    List<HostSpec> currentHosts = this.topologyMap.get(this.clusterId);
+    List<HostSpec> latestHosts;
+
+    synchronized (this.requestToUpdateTopology) {
+      this.requestToUpdateTopology.set(true);
+
+      // Notify monitoring thread (that might be sleeping) that topology should be refreshed immediately.
+      this.requestToUpdateTopology.notifyAll();
+    }
+
+    if (timeoutMs == 0) {
+      return currentHosts;
```

Comment:
Passing a 0 timeout was an intention. In some cases, a plugin doesn't need to wait for an updated topology since the current (a copy from cache) already "good enough". For example, during a reader failover there're few available readers to reconnect. An up-to-date writer isn't important in this case.

Reply:
Right, my point was that the convention we used in RdsHostListProvider is that `refresh()` means "grab me any topology info, I don't care if its cached or not" and `forceRefresh()` specifically means "grab me new topology info, not cached topology info". So it seems unexpected that `forceRefresh()` would return cached topology here

Reply:
I see. But this method actually forces topology to update. It doesn't wait for results.

Reply:
Right. As long as the caller doesn't expect updated topology when they pass in a timeout value of 0 this should be fine. Can we maybe document that passing a value of 0 forces a topology update but also returns cached topology? Since, in some contexts, passing a value of 0 means "no timeout, wait however long it take".